src/libsystemd-network/dhcp-option.c

```diff
@@ -93,32 +112,74 @@ static int option_append(uint8_t options[], size_t size, size_t *offset,
 
                 options[*offset] = code;
                 options[*offset + 1] = l;
-
                 *offset += 2;
 
                 ORDERED_SET_FOREACH(p, s) {
-                        options[*offset] = p->option;
-                        options[*offset + 1] = p->length;
-                        memcpy(&options[*offset + 2], p->data, p->length);
-                        *offset += 2 + p->length;
+                        assert_se(dhcp_option_append_tlv(options, size, offset, p->option, p->length, p->data) >= 0);
+                }
+                break;
+        }
+        case SD_DHCP_OPTION_RELAY_AGENT_INFORMATION: {
+                sd_dhcp_server *server = (sd_dhcp_server *) optval;
+                size_t current_offset = *offset + 2;
+
+                if (server->agent_circuit_id) {
+                        size_t length = strlen(server->agent_circuit_id);
+                        int subcode = SD_DHCP_RELAY_AGENT_CIRCUIT_ID;
+                        r = dhcp_option_append_tlv(options, size, &current_offset, subcode, length, server->agent_circuit_id);
+                        if (r < 0)
+                                return r;
+                }
+                if (server->agent_remote_id) {
+                        size_t length = strlen(server->agent_remote_id);
+                        int subcode = SD_DHCP_RELAY_AGENT_REMOTE_ID;
+                        r = dhcp_option_append_tlv(options, size, &current_offset, subcode, length, server->agent_remote_id);
+                        if (r < 0)
+                                return r;
                 }
 
+                options[*offset] = code;
+                options[*offset + 1] = current_offset - *offset;
+                assert(options[*offset + 1] <= 255);
```

Comment:
The length is not checked anywhere, so setting long IDs triggers this assertion.

Reply:
it's checked in sd_dhcp_server_set_relay_agent_circuit_id and sd_dhcp_server_set_agent_remote_id
PS assert is wrong, option is an char array, fixed this